bundles/org.openhab.binding.metofficedatahub/src/main/java/org/openhab/binding/metofficedatahub/internal/MetOfficeDataHubBridgeHandler.java

```diff
@@ -0,0 +1,174 @@
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link MetOfficeDataHubBridgeHandler} models the account(s) to the MetOfficeDataHub services.
+ *
+ * @author David Goodyear - Initial contribution
+ */
+@NonNullByDefault
+public class MetOfficeDataHubBridgeHandler extends BaseBridgeHandler {
+
+    private volatile MetOfficeDataHubBridgeConfiguration config = getConfigAs(
+            MetOfficeDataHubBridgeConfiguration.class);
+
+    private final TranslationProvider translationProvider;
+    private final LocaleProvider localeProvider;
+    private final Bundle bundle;
+    private final HttpClient httpClient;
+
+    private final Logger logger = LoggerFactory.getLogger(MetOfficeDataHubBridgeHandler.class);
+    private final Object timerResetSchedulerLock = new Object();
+
+    private @Nullable ScheduledFuture<?> timerResetScheduler = null;
+    private @Nullable ScheduledFuture initTask;
+
+    protected final RequestLimiter forecastDataLimiter = new RequestLimiter();
+
+    public MetOfficeDataHubBridgeHandler(final Bridge bridge, IHttpClientProvider httpClientProvider,
+            @Reference TranslationProvider translationProvider, @Reference LocaleProvider localeProvider) {
+        super(bridge);
+        this.translationProvider = translationProvider;
+        this.localeProvider = localeProvider;
+        this.bundle = FrameworkUtil.getBundle(getClass());
+        this.httpClient = httpClientProvider.getHttpClient();
+    }
+
+    protected String getApiKey() {
+        return config.siteApiKey;
+    }
+
+    public String getLocalizedText(String key, @Nullable Object @Nullable... arguments) {
+        String result = translationProvider.getText(bundle, key, key, localeProvider.getLocale(), arguments);
+        return Objects.nonNull(result) ? result : key;
+    }
+
+    public void updateLimiterStats() {
+        final Map<String, String> newProps = new HashMap<>();
+        newProps.put(BRIDGE_PROP_FORECAST_REQUEST_COUNT, String.valueOf(forecastDataLimiter.getCurrentRequestCount()));
+        this.updateProperties(newProps);
+    }
+
+    private static long getMillisUntilMidnight() {
+        return Duration.between(LocalDateTime.now(), LocalDate.now().plusDays(1).atStartOfDay()).toMillis();
+    }
+
+    protected static long getMillisSinceDayStart() {
+        return Duration.between(LocalDate.now().atStartOfDay(), LocalDateTime.now()).toMillis();
+    }
+
+    @Override
+    public void initialize() {
+        updateLimiterStats();
+        config = getConfigAs(MetOfficeDataHubBridgeConfiguration.class);
+        forecastDataLimiter.updateLimit(config.siteRateDailyLimit);
+
+        updateStatus(ThingStatus.UNKNOWN);
+        scheduleResetDailyLimiters();
+
+        initTask = scheduler.schedule(() -> {
+            final PointType siteApiTestLocation = new PointType("51.5072,0.1276");
+            updateLimiterStats();
```

Comment:
initialize() starts with a `updateLimiterStats();` call, then `scheduleResetDailyLimiters();` and async it calls another `updateLimiterStats();` In between also a call to `forecastDataLimiter.updateLimit(config.siteRateDailyLimit);` has been made.
This seems inconvenient and error prone. I would suggest to:

1. Any change to the count in `forecastDataLimiter` should lead to an update of the properties. Maybe you can add a listener or callback pattern?
2. This `initialize()` should reset the `forecastDataLimiter` as history would become erased. According to 1. this can be one call to start with.

Reply:
Ironically I hadn't seen this before, but I have attached a listener approach to make the system more async, In regards to the data limiter that now uses persistent storage so hopefully it should restore with minimal error the last known value.